CASMPET-7252: Update tapms documentation (#5448)
Cray-HPE/docs-csm: operations/multi-tenancy/ExampleWorkflow.md

* [Appendices](#appendices)
    * [Appendix A - `Noname` tenant](#appendix-a---noname-tenant)
    * [Appendix B - `Noname Slurmcluster`](#appendix-b---noname-slurmcluster)
    * [Appendix C - Slurm configuration](#appendix-c---slurm-configuration)
    * [Appendix D - USS group variables](#appendix-d---uss-group-variables)

## Overview

A tenant is a collection of nodes that is dedicated to one particular set of users on an HPE Cray EX computer system running CSM.
This guide is intended to provide a comprehensive set of instructions for a system administrator to configure, deploy, and run applications on, one or two tenants.

In this document we provide examples for a hypothetical system called `Noname`, which has two tenants, and each tenant has a `Slurmcluster`.

Please note that this document reflects the current state of the Multi-Tenancy feature.  For example, VNI blocks must be manually configured today, but they will be automatically configured in a future release.

Here are the steps required:

* Configure desired tenants and `Slurmclusters`
* Deploy each tenant and its `Slurmcluster`
* Configure each `Slurmcluster`'s `slurm.conf`

> * Configure the primary (`user` namespace) `slurm.conf`

Reviewer: I now think this woud be better:  Make any required changes (e.g. VNIs) to primary (user namespace) slurm.conf

Author: updated

    ```

### USS Configuration

These changes are made to the `uss-config-management` git repo.
Tenants are disambiguated by their HSM group name (e.g. `hsmgrouplabel` `"noten01a"` in `noten01a.yaml`, and `hsmGroup` `"noten01a"` in `nocls01a.yaml`).
All tenants can be booted and configured with a single CFS configuration that contains the appropriate git commit ID in the USS layers.

* Example: `group_vars/noten01a/slurm.yml`

    ```yaml
    munge_vault_path: secret/slurm/vcluster-noten01a-slurm/nocls01a/munge
    slurm_conf_url: https://rgw-vip.local/wlm/vcluster-noten05-slurm/nocls01a/
    slurmd_options: "--conf-server 10.156.124.104,10.156.124.105"
    ```

* Example:  `group_vars/noten02a/slurm.yml`

    ```yaml
    munge_vault_path: secret/slurm/vcluster-noten02a-slurm/nocls02a/munge
    slurm_conf_url: https://rgw-vip.local/wlm/vcluster-noten02a-slurm/nocls02a/
    slurmd_options: "--conf-server 10.156.124.108,10.156.124.109"
    ```

## Step-by-Step Guide

Legend for these examples:

* `noten01a.yaml` - configuration file for tenant `noten01a`
* `nocls01a.yaml` - configuration file for `Slurmcluster` `nocls01a`

> ### Create tenant configuration file

Reviewer: Create and apply tenant configuration file

Author: fixed

Filename:  `noten01a.yaml`

* (`ncn-mw#`) Make sure tenant name is not already in use:

    ```bash
    kubectl get tenant -n tenants -o yaml vcluster-noten01a
    ```

* (`ncn-mw#`) Make sure HSM group is not already in use:

    ```bash
    cray hsm groups describe noten01a
    ```

* (`ncn-mw#`) Create your `tenant.yaml` file, and apply it:
    * See Appendix A for an example

    ```bash
    vi noten01a.yaml
    kubectl apply -n tenants -f noten01a.yaml
    ```

* (`ncn-mw#`) Wait for 'Deploying' state to become 'Deployed':

    ```bash
    kubectl get tenant -n tenants -o yaml vcluster-noten01a
    ```

* (`ncn-mw#`) Confirm HSM group:

    ```bash
    cray hsm groups describe noten01a
    ```

Repeat this step as needed for additional tenants.

> ### Create `Slurmcluster` configuration file

Reviewer: Create and apply Slurmcluster configuration file

Author: fixed

Filename:  `nocls01a.yaml`

* (`ncn-mw#`) Make sure cluster name is not already in use:

    ```bash
    kubectl get pods -A | grep vcluster-nocls01a-slurm
    ```

* (`ncn-mw#`) Create your `cluster.yaml` file, and apply it:
    * See Appendix B for an example

    ```bash
    vi nocls01a.yaml
    kubectl apply -f nocls01a.yaml
    ```

* (`ncn-mw#`) Wait for pods to initialize:

    ```bash
    kubectl get pods -A | grep vcluster-noten01a-slurm
    ```

Repeat this step as needed for additional `Slurmclusters`.

> ### Edit Slurm configuration file

Reviewer: Edit and apply Slurm configuration file

Author: fixed

    ```bash
    git commit -am "descriptive comment"
    ```

* (`ncn-mw#`) Push to integration branch (1.1.0 shown here):

    ```bash
    git push origin integration-1.1.0
    ```

* (`ncn-mw#`) Remember the first commit ID in the output:

    ```bash
    git log -a |cat
    ```

    Repeat this step as needed for additional tenants.

### Create BOS session template

Note that you will need one template for each node type (UAN,Compute) and architecture (X86,ARM) in the tenants.
You can use a single BOS session template for many tenants of the same node type and architecture.

* (`ncn-mw#`) Look up the name of the default template(s) for tenants (e.g. X86 Compute) and save as JSON file(s):

    ```bash
    cray bos sessiontemplates describe --format json ssi-compute-cr_2024.x86_64-cr_2024_1 > ssi-compute-cr_2024.x86_64-cr_2024_1.json
    ```

* (`ncn-mw#`) Make a copy of the default template:

    ```bash
    cp ssi-compute-cr_2024.x86_64-cr_2024_1.json ssi-compute-cr_2024.x86_64-cr_2024_1-tenants.json
    ```

* Edit the new copy:
    * Delete block of lines starting with `enable_cfs:`, `name:`, and `tenant:` and remove the comma from the preceding line

>     * Change the name of the CFS configuration to include the tenant name

Reviewer: Change the name of the CFS configuration in the BOS session template to add a '-tenants' suffix, as seen in the next section

Author: fixed

* (`ncn-mw#`) Upload the new template, specifying the filename and the name of the new template:

    ```bash
    cray bos sessiontemplates create --format json --file ssi-compute-cr_2024.x86_64-cr_2024_1-tenants.json ssi-compute-cr_2024.x86_64-cr_2024_1-tenants
    ```

Repeat this step as needed for different node types and architectures.

### Create CFS configuration

Note that you will need one configuration for each node type (UAN,Compute) in the tenant.
You can use a single BOS session template for many tenants of the same node type and architecture.

* (`ncn-mw#`) Save the default configuration as a JSON file:

    ```bash
    cray cfs configurations describe --format json ssi-compute-cr_2024-cr_2024_1 > ssi-compute-cr_2024-cr_2024_1.json
    ```

* (`ncn-mw#`) Make a copy of the default JSON file:

    ```bash

>     cp ssi-compute-cr_2024-cr_2024_1.json ssi-compute-cr_2024-cr_2024_1-noten01a.json

Reviewer: cp ssi-compute-cr_2024-cr_2024_1.json ssi-compute-cr_2024-cr_2024_1-tenants.json

Author: fixed

    ```

* Edit the new copy (see appendix F for an example):
    * Delete lines starting with `lastUpdated`:
    * Delete the last instance in the file of name: and remove the comma from the preceding line

>     * Be SURE to replace the commit ID for each JSON block that refers to `uss-config-management.git`; you will use the commit ID from "git log" command in the earlier step to create the USS `group_vars` file

Reviewer: End of sentence:  ... in the earlier step that created the USS group_vars file

Author: fixed

* (`ncn-mw#`) Upload the new configuration, specifying the filename and the name of the new configuration:

    ```bash
    # cray cfs configurations update --file ssi-compute-cr_2024-cr_2024_1-tenants ssi-compute-cr_2024-cr_2024_1-tenants.json
    ```

Repeat this step as needed for different node types and architectures.

### Boot and Run

* BOS and CFS:
    * You'll need one or more BOS session template(s) for your tenant, see above
    * The BOS session template(s) refer to CFS configuration(s), see above

* (`ncn-mw#`) Boot Compute nodes for a node architecture in tenant:

    ```bash

>     cray bos sessions create --template-name ssi-compute-cr_2024.x86_64-cr_2024_1-noten01a --operation boot --limit x9000c1s0b1n0,x9000c1s0b1n1

Reviewer: cray bos sessions create --template-name ssi-compute-cr_2024.x86_64-cr_2024_1-tenants --operation boot --limit x9000c1s0b1n0,x9000c1s0b1n1

Author: fixed

    * [Create `Slurmcluster` configuration file](#create-and-apply-slurmcluster-configuration-file)
    * [Edit Slurm configuration file](#edit-and-apply-slurm-configuration-file)
    * [Edit Slurm configuration file `sssd.conf`](#edit-slurm-configuration-file-sssdconf)
    * [Configure USS group variables](#configure-uss-group-variables)
    * [Create BOS session template](#create-bos-session-template)
    * [Create CFS configuration](#create-cfs-configuration)
    * [Boot and Run](#boot-and-run)
* [Status and Troubleshooting](#status-and-troubleshooting)
    * [Tenant command examples](#tenant-command-examples)
    * [`Slurmcluster` command examples](#slurmcluster-command-examples)
    * [HSM command examples](#hsm-command-examples)
    * [HNS command example](#hns-command-example)
* [Appendices](#appendices)
    * [Appendix A - `Noname` tenant](#appendix-a---noname-tenant)
    * [Appendix B - `Noname Slurmcluster`](#appendix-b---noname-slurmcluster)
    * [Appendix C - Slurm configuration](#appendix-c---slurm-configuration)
    * [Appendix D - USS group variables](#appendix-d---uss-group-variables)

## Overview

> A tenant is a collection of nodes that is dedicated to one particular set of users on an HPE Cray EX computer system running CSM.

Reviewer 2: ```suggestion
A tenant is a collection of nodes that is dedicated to one particular set of users on an HPE Cray EX system running CSM.
```

This guide is intended to provide a comprehensive set of instructions for a system administrator to configure, deploy, and run applications on, one or two tenants.

> In this document we provide examples for a hypothetical system called `Noname`, which has two tenants, and each tenant has a `Slurmcluster`.

Reviewer 2: Is noname an internal system name? We might want to change this name if it is since this is a public repository.

Author: I'm not sure why this name was chosen since I'm going off of the draft someone else gave me.  I've changed this to "Development" to be safe.

> Please note that this document reflects the current state of the Multi-Tenancy feature.  For example, VNI blocks must be manually configured today, but they will be automatically configured in a future release.

Reviewer 2: ```suggestion
Note that this document reflects the current state of the Multi-Tenancy feature.  For example, VNI blocks must be manually configured today, but they will be automatically configured in a future release.
```

Author: fixed

* **Name length limitation**:
    * longer names are automatically generated from `Slurmcluster` name
    * for example, `nocls01a-slurmdb` (16 characters)
    * name length limitation is 22 characters

* Example of settings in configuration file `nocls01a.yaml`:

    ```yaml
    namespace: vcluster-noten01a-slurm
    tapmsTenantName: vcluster-noten01a
    hsmGroup: noten01a
    ```

#### `Slurmcluster` IP Addresses

**`IMPORTANT`** Each High-Speed Network (HSN) IP address must be unique, within all the `Slurmclusters` on any one system.

* These HSN IP addresses are assigned in the USS configuration, below
    * **You will need to know the base HSN IP address for each system**
* Four HSN IP addresses are used in each `Slurmcluster`
* Example:
    * System `Noname` base HSN IP address 10.156.0.0
    * Primary `Slurmcluster` (`user` namespace) uses 10.156.12.100, .101, .102, .103
    * First tenant `Slurmcluster` (`vcluster-noten01-slurmdb` namespace) will use 10.156.12.104, .105, .106, .107
    * Second tenant `Slurmcluster` (`vcluster-noten02-slurmdb` namespace) will use 10.156.12.108, .109, .110, .111

#### `Slurmcluster` API version

> * `Slurmcluster` `apiVersion` must match Slurm release (e.g. `v1alpha1`)

Reviewer 2: ```suggestion
* `Slurmcluster` `apiVersion` must match Slurm release (for example, `v1alpha1`)
```

#### `Slurmcluster` Configurable Values

* Settings for `cpu` and `memory` and `initialDelaySeconds` are shown in the example file `nocls01a.yaml`, below
* These settings were provided by the WLM team, who should be consulted for any changes

#### `Slurmcluster` Version Numbers

* Version numbers are shown in the example file `nocls01a.yaml`, below
* The version numbers must match the versions of these products on the system

* Example:

    ```yaml
    - cray/cray-slurmctld:1.6.1
    - cray/cray-slurmdbd:1.6.1
    - cray/munge-munge:1.5.0
    - cray/cray-sssd:1.4.0
    - cray/cray-slurm-config:1.3.0
    ```

### `Slurmcluster` Slurm configuration

These configuration settings are made in each `Slurmcluster's` `/etc/slurm/slurm.conf` file (in each `slurmctld` pod)

#### `Slurmcluster` Slingshot VNI Allocation

> **`IMPORTANT`** Each block of Slingshot `VNIs` on the High-Speed Network (HSN) must not overlap with other blocks on the same system.

Reviewer 2: ```suggestion
**`IMPORTANT`** Each block of HPE Slingshot `VNIs` on the High-Speed Network (HSN) must not overlap with other blocks on the same system.
```

        * `SwitchParameters=vnis=57354-65535`

#### `Slurmcluster` Partitions and Nodes

The general advice to tailor the compute node configuration for each tenant is to look at the `slurm.conf` for the primary (`user` namespace) Slurm instance.
Borrow the `NodeSet`, `PartitionName`, and `NodeName` directives that apply to each tenant.

* In this example, we have 'moved' two Compute nodes to the `slurm.conf` in namespace `vcluster-noten01a-slurm`.

    ```ini
    # PARTITIONS
    NodeSet=Compute Feature=Compute
    PartitionName=workq Nodes=Compute MaxTime=INFINITE State=UP OverSubscribe=EXCLUSIVE
    # BEGIN COMPUTE NODES
    NodeName=nid[001002-001003] Sockets=2 CoresPerSocket=64 ThreadsPerCore=2 RealMemory=456704 Feature=Compute
    ```

#### `Slurmcluster` Secrets (for `nonroot` users)

These configuration settings are made in each `Slurmcluster's` `/etc/sssd/sssd.conf` file (in each `slurmctld` pod)
You should not need to create or edit the `sssd.conf` file.
Simply clone that file from the primary `Slurmcluster` (`user` namespace) to each tenant namespace.

### USS Configuration

These changes are made to the `uss-config-management` git repo.

> Tenants are disambiguated by their HSM group name (e.g. `hsmgrouplabel` `"noten01a"` in `noten01a.yaml`, and `hsmGroup` `"noten01a"` in `nocls01a.yaml`).

Reviewer 2: ```suggestion
Tenants are disambiguated by their HSM group name (For example, `hsmgrouplabel` `"noten01a"` in `noten01a.yaml`, and `hsmGroup` `"noten01a"` in `nocls01a.yaml`).
```

All tenants can be booted and configured with a single CFS configuration that contains the appropriate git commit ID in the USS layers.

* Example: `group_vars/noten01a/slurm.yml`

    ```yaml
    munge_vault_path: secret/slurm/vcluster-noten01a-slurm/nocls01a/munge
    slurm_conf_url: https://rgw-vip.local/wlm/vcluster-noten05-slurm/nocls01a/
    slurmd_options: "--conf-server 10.156.124.104,10.156.124.105"
    ```

* Example:  `group_vars/noten02a/slurm.yml`

    ```yaml
    munge_vault_path: secret/slurm/vcluster-noten02a-slurm/nocls02a/munge
    slurm_conf_url: https://rgw-vip.local/wlm/vcluster-noten02a-slurm/nocls02a/
    slurmd_options: "--conf-server 10.156.124.108,10.156.124.109"
    ```

> ## Step-by-Step Guide

Reviewer 2: The following sections should use an ordered list instead of an unordered list since it is sequential steps.

    ```

* (`ncn-mw#`) Wait for 'Deploying' state to become 'Deployed':

    ```bash
    kubectl get tenant -n tenants -o yaml vcluster-noten01a
    ```

* (`ncn-mw#`) Confirm HSM group:

    ```bash
    cray hsm groups describe noten01a
    ```

Repeat this step as needed for additional tenants.

### Create and apply `Slurmcluster` configuration file

Filename:  `nocls01a.yaml`

* (`ncn-mw#`) Make sure cluster name is not already in use:

    ```bash
    kubectl get pods -A | grep vcluster-nocls01a-slurm
    ```

* (`ncn-mw#`) Create your `cluster.yaml` file, and apply it:
    * See Appendix B for an example

    ```bash
    vi nocls01a.yaml
    kubectl apply -f nocls01a.yaml
    ```

* (`ncn-mw#`) Wait for pods to initialize:

    ```bash
    kubectl get pods -A | grep vcluster-noten01a-slurm
    ```

Repeat this step as needed for additional `Slurmclusters`.

### Edit and apply Slurm configuration file

``Slurmcluster``:  `nocls01a`
Filename:  `/etc/slurm/slurm.conf`

* (`ncn-mw#`) Get the running configuration:

    ```bash
    kubectl get configmap -n vcluster-noten01a-slurm nocls01a-slurm-conf -o yaml > nocls01a-slurm-conf.yaml
    ```

> * (`ncn-mw#`) Extract the slurm.conf:

Reviewer 2: ```suggestion
* (`ncn-mw#`) Extract the `slurm.conf`:
```

    ```bash
    yq r nocls01a-slurm-conf.yaml 'data."slurm.conf"' > slurm.conf
    ```

> * (`ncn-mw#`) Edit the slurm.conf:

Reviewer 2: ```suggestion
* (`ncn-mw#`) Edit the `slurm.conf`:
```

    * See Appendix C

    ```bash
    vi slurm.conf
    ```

1) (`ncn-mw#`) Update the configuration:

    ```bash
    yq w -i devcls01a-slurm-conf.yaml 'data."slurm.conf"' "$(cat slurm.conf)"
    ```

1) (`ncn-mw#`) Apply the configuration:

    ```bash
    kubectl apply -f devcls01a-slurm-conf.yaml
    ```

1) (`ncn-mw#`) Look up the pod for the tenant `slurmcluster`:

    ```bash
    SLURMCTLD_POD=$(kubectl get pod -n vcluster-devten01a-slurm -lapp=slurmctld -o name)
    ```

1) (`ncn-mw#`) Reconfigure:

    ```bash
    kubectl exec -n user ${SLURMCTLD_POD} -c slurmctld -- scontrol reconfigure
    ```

Repeat this step as needed for additional `Slurmclusters`.

### Edit Slurm configuration file `sssd.conf`

`Slurmcluster`:  `devcls01a`
Filename:  `/etc/sssd/sssd.conf`

> 1) (`ncn-mw#`) Get the `user` namespace `sssd.conf` so it can be cloned:

Reviewer 2: These should be formatted as `1.` so they are rendered properly

    ```

* (`ncn-mw#`) Go to repo:

    ```bash
    cd uss-config-management
    ```

* (`ncn-mw#`) Check out integration branch (1.1.0 shown here):

    ```bash
    git checkout integration-1.1.0
    ```

* (`ncn-mw#`) Create subdirectory for tenant:

    ```bash
    mkdir group_vars/noten01a
    ```

* (`ncn-mw#`) Edit the file `group_vars/noten01a/slurm.yml`
    * See an appropriate example in Appendix D

* (`ncn-mw#`) Add the new file:

    ```bash
    git add group_vars/noten01a/slurm.yml
    ```

* (`ncn-mw#`) Commit the new file:

    ```bash
    git commit -am "descriptive comment"
    ```

* (`ncn-mw#`) Push to integration branch (1.1.0 shown here):

    ```bash
    git push origin integration-1.1.0
    ```

* (`ncn-mw#`) Remember the first commit ID in the output:

    ```bash
    git log -a |cat
    ```

    Repeat this step as needed for additional tenants.

### Create BOS session template

Note that you will need one template for each node type (UAN,Compute) and architecture (X86,ARM) in the tenants.
You can use a single BOS session template for many tenants of the same node type and architecture.

> * (`ncn-mw#`) Look up the name of the default template(s) for tenants (e.g. X86 Compute) and save as JSON file(s):

Reviewer 2: ```suggestion
* (`ncn-mw#`) Look up the name of the default template(s) for tenants (for example, X86 Compute) and save as JSON file(s):
```

* (`ncn-mw#`) Save the default configuration as a JSON file:

    ```bash
    cray cfs configurations describe --format json ssi-compute-cr_2024-cr_2024_1 > ssi-compute-cr_2024-cr_2024_1.json
    ```

* (`ncn-mw#`) Make a copy of the default JSON file:

    ```bash
    cp ssi-compute-cr_2024-cr_2024_1.json ssi-compute-cr_2024-cr_2024_1-tenants.json
    ```

* Edit the new copy (see appendix F for an example):
    * Delete lines starting with `lastUpdated`:
    * Delete the last instance in the file of name: and remove the comma from the preceding line
    * Be SURE to replace the commit ID for each JSON block that refers to `uss-config-management.git`; you will use the commit ID from "git log" command in the earlier step that created the USS `group_vars` file

* (`ncn-mw#`) Upload the new configuration, specifying the filename and the name of the new configuration:

    ```bash
    # cray cfs configurations update --file ssi-compute-cr_2024-cr_2024_1-tenants ssi-compute-cr_2024-cr_2024_1-tenants.json
    ```

Repeat this step as needed for different node types and architectures.

### Boot and Run

* BOS and CFS:

>     * You'll need one or more BOS session template(s) for your tenant, see above

Reviewer 2: ```suggestion
    * You will need one or more BOS session template(s) for your tenant, see above
```

    * The BOS session template(s) refer to CFS configuration(s), see above

* (`ncn-mw#`) Boot Compute nodes for a node architecture in tenant:

    ```bash
    cray bos sessions create --template-name ssi-compute-cr_2024.x86_64-cr_2024_1-tenants --operation boot --limit x9000c1s0b1n0,x9000c1s0b1n1
    ```

After CFS completes, login to either a tenant UAN (if available), or tenant Compute.

* (`ncn-mw#`) See what nodes are available:

    ```bash
    sinfo
    ```

* (`ncn-mw#`) Launch a command or application:

    ```bash
    srun -N2 uname -rin
    srun -N2 ./all2all
    ```

### Status and Troubleshooting

#### Tenant command examples

> * (`ncn-mw#`) Specific tenant, brief:

Reviewer 2: ```suggestion
* (`ncn-mw#`) View a specific tenant, brief:
```

    ```bash
    kubectl get tenant -n tenants -o yaml vcluster-noten01a
    ```

> * (`ncn-mw#`) Specific tenant, verbose:

Reviewer 2: ```suggestion
* (`ncn-mw#`) View a specific tenant, verbose:
```

    ```bash
    kubectl describe tenant -n tenants vcluster-noten01a
    ```

> * (`ncn-mw#`) All tenants, logs:

Reviewer 2: ```suggestion
* (`ncn-mw#`) View the logs for all tenants.
```

    ```bash
    # TAPMS_POD=$( kubectl get pods -n tapms-operator --no-headers | awk '{print $1}' );
    # kubectl logs --timestamps -n tapms-operator $TAPMS_POD
    ```

#### `Slurmcluster` command examples

> * (`ncn-mw#`) All clusters:

Reviewer 2: ```suggestion
* (`ncn-mw#`) View the pods for all clusters:
```

    ```bash
    kubectl get pods -A | grep vcluster
    ```

> * (`ncn-mw#`) Specific cluster:

Reviewer 2: ```suggestion
* (`ncn-mw#`) View the pods for a specific cluster:
```

    ```bash
    kubectl get pods -A | grep vcluster-noten01a-slurm
    ```

> * (`ncn-mw#`) Specific cluster, logs:

Reviewer 2: ```suggestion
* (`ncn-mw#`) View logs for a specific cluster:
```
